### ISSUE TITLE
Root privilege escalation by subverting chown

### ISSUE THREAD

**Issue by orlitzky** [ltb-project/openldap-initscript#2]
Hi, in the init script, you make use of `chown` in a few places. These calls are made as root, repeatedly, and often automatically (like when the machine reboots). For example,

```sh
piddir=`dirname "$SLAPD_PID_FILE"`
if [ ! -d "$piddir" ]; then
	mkdir -p "$piddir"
	[ -z "$SLAPD_USER" ] || chown -R "$SLAPD_USER" "$piddir"
	[ -z "$SLAPD_GROUP" ] || chgrp -R "$SLAPD_GROUP" "$piddir"
fi
```

There is an exploit here that lets `$SLAPD_USER` gain root on the machine. Above, he is given control of the `$piddir` and everything in it via `chown -R`. Now that he owns that directory, suppose he places a hardlink in there pointing to `/root/.bashrc`. The next time the slapd init script runs, it will call `chown -R` again, and follow the link to give ownership of `/root/.bashrc` to the `$SLAPD_USER`. Voila, he's root.

I found a bunch of similar vulnerabilities last year and wrote an article about it:

http://michael.orlitzky.com/articles/end_root_chowning_now_%28make_etc-init.d_great_again%29.xhtml

tl;dr running `chown` as root (in an automated fashion) is next to impossible to get right.

**Comment by coudot** [comment]
Ok, but not sure there is a security risk as $SLAPD_USER is a systemuser without shell. And in the example you give, the chown is done only at $piddir creation, so it is empty.

But anyway, putting more security in the script is always a good thing. Feel free to propose a patch if you can.

**Comment by orlitzky** [comment]
Ha, no, you are right. I didn't read carefully enough, and didn't even see the `! -d` test. That invocation is probably safe -- there's a race condition there, but so long as `$SLAPD_USER` doesn't own the parent directory I don't think there's any way for him to trick that `! -d` test into passing.

Let me try to contribute something actually useful then. The default backup path is,

```sh
BACKUP_PATH="/tmp/openldap"
```

Later, we have,

```sh
file="$BACKUP_PATH/$dir-$BACKUP_SUFFIX"
...
$SU "$SLAPCAT_BIN -b \"$sufprint\" $SLAPCAT_PARAMS" > "$file"
ret=$?
chown $SLAPD_USER:$SLAPD_GROUP $file
```

If the `$SLAPD_USER` owns `$BACKUP_PATH`, then he can create `$file` as a symlink to something sensitive before the backup script runs. The slapcat command will fail, but chown will still be called on the symlink, allowing `$SLAPD_USER` to steal root's files.

How would `$SLAPD_USER` own `$BACKUP_PATH`? Well, `/tmp` is writable by anyone. So long as `/tmp/openldap` doesn't already exist, he can just create it himself and will wind up as the owner. For example, on my machine,

```sh
$ mkdir /tmp/openldap
$ ls /tmp/ | grep openldap
drwxr-xr-x 2 mjo      mjo       4.0K 2017-09-06 07:53 openldap
```

Fortunately, in this case, I think that `$file` will *already* be created as the `$SLAPD_USER`, thanks to the "su" command. In other words, I bet you can just get rid of the `chown`.

**Comment by davidcoutadeur** [comment]
Indeed, the chown seems useless and has been removed. (see 70643e4)
